sklearn/model_selection/_search_successive_halving.py

```diff
@@ -0,0 +1,895 @@
+                self.min_resources_,
+                self.max_resources_ // self.ratio**last_iteration
+            )
+
+        # n_possible_iterations is the number of iterations that we can
+        # actually do starting from min_resources and without exceeding the
+        # max_resources. Depending on max_resources and the number of
+        # candidates, this may be higher or smaller than
+        # n_required_iterations.
+        n_possible_iterations = 1 + floor(log(
+            self.max_resources_ // self.min_resources_, self.ratio))
+
+        if self.aggressive_elimination:
+            n_iterations = n_required_iterations
+        else:
+            n_iterations = min(n_possible_iterations, n_required_iterations)
+
+        if self.verbose:
+            print('n_iterations: {}'.format(n_iterations))
+            print('n_required_iterations: {}'.format(n_required_iterations))
+            print('n_possible_iterations: {}'.format(n_possible_iterations))
+            print('min_resources_: {}'.format(self.min_resources_))
+            print('max_resources_: {}'.format(self.max_resources_))
+            print('aggressive_elimination: {}'.format(
+                self.aggressive_elimination))
+            print('force_exhaust_resources: {}'.format(
+                self.force_exhaust_resources))
+            print('ratio: {}'.format(self.ratio))
+
+        # list of resource_iter for each iteration, used in tests
+        self._r_i_list = []
+        self.n_candidates_ = []
+
+        for iter_i in range(n_iterations):
+
+            power = iter_i  # default
+            if self.aggressive_elimination:
+                # this will set resource_iter to the initial value (i.e. the
+                # value of resource_iter at the first iteration) for as many
+                # iterations as needed (while candidates are being
+                # eliminated), and then go on as usual.
+                power = max(
+                    0,
+                    iter_i - n_required_iterations + n_possible_iterations
+                )
+
+            resource_iter = int(self.ratio**power * self.min_resources_)
+            # guard, probably not needed
+            resource_iter = min(resource_iter, self.max_resources_)
+            self._r_i_list.append(resource_iter)
+
+            n_candidates = len(candidate_params)
+            self.n_candidates_.append(n_candidates)
+
+            if self.verbose:
+                print('-' * 10)
+                print('iter_i: {}'.format(iter_i))
+                print('n_candidates: {}'.format(n_candidates))
+                print('resource_iter: {}'.format(resource_iter))
+
+            if self.resource == 'n_samples':
+                # Subsample X and y as well as fit_params
+                stratify = y if is_classifier(self.estimator) else None
+                fit_params = OrderedDict(fit_params)
+                X_iter, y_iter, *fit_params_iter_list = resample(
```

Comment:
Side note, this assumes that the `fit_params` are sample-aligned and will potentially break third-party estimators that do not respect our current convention

Reply:
gridsearch only slices the parameters which are sample aligned, and passes along the rest unsliced. Do we want to do the same here?

Reply:
yes! I wasn't aware of `_check_fit_params`, thanks